File: app/src/main/java/com/sakuna63/tumbin/application/activity/HomeActivity.java
package com.sakuna63.tumbin.application.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.FragmentManager;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.f2prateek.dart.HensonNavigable;
import com.sakuna63.tumbin.R;
import com.sakuna63.tumbin.application.contract.PostsContract;
import com.sakuna63.tumbin.application.contract.presenter.DashboardPresenter;
import com.sakuna63.tumbin.application.di.component.ActivityComponent;
import com.sakuna63.tumbin.application.di.component.DaggerHomeComponent;
import com.sakuna63.tumbin.application.di.component.HomeComponent;
import com.sakuna63.tumbin.application.di.module.PostsPresenterModule;
import com.sakuna63.tumbin.application.fragment.PostsFragment;
import com.sakuna63.tumbin.application.fragment.PostsFragmentBuilder;
import com.sakuna63.tumbin.application.misc.AccountManager;
import com.sakuna63.tumbin.application.util.ActivityUtils;

import javax.inject.Inject;

@HensonNavigable
public class HomeActivity extends BaseActivity {
    @Inject
    DashboardPresenter presenter;

    @Inject
    AccountManager accountManager;

    private HomeComponent component;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_home);

        Toolbar toolBar = (Toolbar) findViewById(R.id.toolbar);
        if (toolBar != null) {
            setSupportActionBar(toolBar);
        }

        FragmentManager fm = getSupportFragmentManager();
        PostsFragment fragment = (PostsFragment) fm.findFragmentByTag(PostsFragment.TAG);
        if (fragment == null) {
            fragment = new PostsFragmentBuilder().build();
            ActivityUtils.addFragment(fm, fragment, R.id.container_fragment, PostsFragment.TAG);
        }

        initInjector(fragment);

        if (!accountManager.isLoggedIn()) {
            Intent intent = Henson.with(this).gotoLoginActivity().build();
            startActivity(intent);
            finish();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        presenter.destroy();
    }

    @Override
    public ActivityComponent getActivityComponent() {
        return component;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_home, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_reload:
                presenter.onReloadClick();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void initInjector(PostsContract.View view) {
        component = DaggerHomeComponent.builder()
                .applicationComponent(getApplicationComponent())
                .activityModule(getActivityModule())
                .postsPresenterModule(new PostsPresenterModule(view))
                .build();
        component.inject(this);
    }
}

File: app/src/main/java/com/sakuna63/tumbin/application/activity/HomeActivity.kt
package com.sakuna63.tumbin.application.activity

import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import com.sakuna63.tumbin.R
import com.sakuna63.tumbin.application.contract.PostsContract
import com.sakuna63.tumbin.application.contract.presenter.DashboardPresenter
import com.sakuna63.tumbin.application.di.component.ActivityComponent
import com.sakuna63.tumbin.application.di.component.HomeComponent
import com.sakuna63.tumbin.application.di.module.PostsPresenterModule
import com.sakuna63.tumbin.application.fragment.PostsFragment
import com.sakuna63.tumbin.application.misc.AccountManager
import com.sakuna63.tumbin.extension.addFragment
import com.sakuna63.tumbin.extension.bindView
import javax.inject.Inject

class HomeActivity : BaseActivity() {
    @Inject
    lateinit internal var presenter: DashboardPresenter

    @Inject
    lateinit internal var accountManager: AccountManager

    private val toolBar: Toolbar by bindView(R.id.toolbar)

    lateinit private var component: HomeComponent

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        setSupportActionBar(toolBar)

        val fm = supportFragmentManager
        var fragment: PostsFragment? = fm.findFragmentByTag(PostsFragment.TAG) as PostsFragment?
        if (fragment == null) {
            fragment = PostsFragment.newInstance()
            addFragment(fragment, R.id.container_fragment, PostsFragment.TAG)
        }

        initInjector(fragment)

        if (!accountManager.isLoggedIn) {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        presenter.destroy()
    }

    override val activityComponent: ActivityComponent
        get() = component

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_home, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_reload -> {
                presenter.onReloadClick()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun initInjector(view: PostsContract.View) {
        component = applicationComponent.plus(activityModule, PostsPresenterModule(view))
        component.inject(this)
    }
}
